Add support for arbitrary PM constraint lists from DT and substate change notifier, refactor RW612 power domain

decsny opened the pull request:
Cherry-picked the RW612 DTS refactor from https://github.com/zephyrproject-rtos/zephyr/pull/89372 and realized it causes runtime failures due to for example the flexcomm driver is trying to use the `pm_policy_device_is_disabling_state` API which stops working if the disabling state is not directly defined on the device's node and is defined on the parent instead,  so fix this in the policy_device_lock.c by checking first parent domain constraints

bjarki-andreasen-nordic committed:
dts: arm: nxp: rw6xx_common.dtsi: use default power domain

Update board to use default power domain for the "peripheral"
domain, which when resumed prevents SoC from entering suspend and
standby states as these power down peripherals.

Note by co-author: Cherry-picked this commit and edited it to work
without the generic domain refactoring on the branch it came from.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>
Co-authored-by: Declan Snyder <declan.snyder@nxp.com>

bjarki-andreasen reviewed (commented):
Child devices will resume their power domain, and the power domain locks the state. In which scenarios would the child device lock a state of another device? seems like a strong layering violation to have devices look into implementation details of other devices instead of just resuming them, which causes other devices to lock their states already (or do whatever they need to to remain powered)

decsny commented:
> Child devices will resume their power domain, and the power domain locks the state. In which scenarios would the child device lock a state of another device? seems like a strong layering violation to have devices look into implementation details of other devices instead of just resuming them, which causes other devices to lock their states already (or do whatever they need to to remain powered)

You are right, but my intention though was not about actually changing what states are locked, but mostly the issue is related to this [pm_policy_device_lock_is_disabling_state](https://docs.zephyrproject.org/latest/doxygen/html/group__subsys__pm__sys__policy.html#ga4cfc3c450387663b6e1bf9bb966baad8) which now if the DTS is refactored so that the parent domain is has that info, then the function doesnt return the right value (true/false) for the child. 

For example if the parent will be disabled by a mode "suspend", then the child also obviously would be disabled by this mode "suspend" even if `zephyr,disabling-states` is not present on the child but only on the parent. So currently that function which only checks the `zephyr,disabling-states` that is actually on the node for the device in question will return "false" for this child (saying it won't be disabled in "suspend") when in reality it will because it's power domain was disabled. 

And to be clear, this is not changing anything if a device has this property already, only if it doesn't then it checks it's power domain parents based on the assumption that  if the power domain parents are disabled then it will be too.

And beyond just checking if a state will disable a device, as far as the locking mechanism itself - I agree with you that how it's supposed to work is that the power domain parent (in this context) should be the one that locks from entering the state and then the other devices don't need to do this due to as long as one of them is referenced/active then the parent should remain active. I completely understand that and agree. So since my goal wasn't actually to change this locking mechanism, should I change this code to uncouple the pm_policy_device_lock_is_disabling_state to not use the same constraint searching code as the other functions in this file ? I.e, still adding the "special" value of 0 size to indicate to look at the parent device but don't actually do it from the locking functions and only from this checking function (pm_policy_device_lock_is_disabling_state)

decsny commented:
@bjarki-andreasen I think the confusion here is about why is this device needing to check on if it is a disabling state. Because the point is not to support/accommodate system managed device PM. IIRC, the reason is because this device and some others on this chip actually do not lose function in a low power mode, even if the "domain" suspends, they dont turn off and actually continue operating. But the issue with this chip is that there is not really a coherent concept of domains in the hardware like the zephyr architecture is meant for, it's just a bunch of barely documented behaviors that are seemingly different for each peripheral on the chip. Some of them lose state at certain power mode, some of them still function, some of them retain state but dont function.... and it's not configurable or clear why or how (we mostly are figuring it out by reverse engineering)...

So IIRC this UART is one of them that goes in a "low power" mode when system enters this "PM2" mode (associated to suspend in zephyr). In this case, the UART is not able to really do any send and transceive normally as far as I could tell but it could change clock source to some really slow like 1Hz clock which is the only one left active on the system in this mode, and be able to "receive" one bit/byte, and this activity can then be a wakeup cause of the chip. So what this driver is essentially saying is "if my normal function is going to be disabled, reconfigure my clock to the low power clock and configure me as wake up source" (zephyr,disabling-states). Whereas on "PM3" mode (associated to suspend), the UART just completely lose power and state and cannot be a wakeup. So we set the power-domain to this "PM3" domain which is currently just a stupid mechanism to be able to deliver an "OFF" action to all these devices which lose power and state from "standby" mode, and set "zephyr,disabling-states" as "suspend" (and standby technically, but irrelevant) so that it knows it loses function but can reconfigure itself as a wakeup...

So do you have any idea about how to handle this problem which is that our device cannot only just be simply "disabled" or "enabled" but can also be in this type of low power mode that it needs to account for? Because from PM device callbacks alone, it's not enough here to handle that "PM2" case, we can handle this PM3 case to get the OFF action and know we will lose state but I am not sure how to differentiate for the "low power operation" case

decsny commented:
And by the way, I am noticing this is a pretty common thing for us too - for power domains and devices to have not just a binary on/off state but actually different levels of operation (low power modes). Right now we only enabled this uart driver for it's low power mode because we wanted the shell to keep working on PM2 since many of our wifi samples (RW612 is basically mainly a wifi chip) were based on a shell

bjarki-andreasen commented:
> @bjarki-andreasen I think the confusion here is about why is this device needing to check on if it is a disabling state. Because the point is not to support/accommodate system managed device PM. IIRC, the reason is because this device and some others on this chip actually do not lose function in a low power mode, even if the "domain" suspends, they dont turn off and actually continue operating. But the issue with this chip is that there is not really a coherent concept of domains in the hardware like the zephyr architecture is meant for, it's just a bunch of barely documented behaviors that are seemingly different for each peripheral on the chip. Some of them lose state at certain power mode, some of them still function, some of them retain state but dont function.... and it's not configurable or clear why or how (we mostly are figuring it out by reverse engineering)...
> 
> So IIRC this UART is one of them that goes in a "low power" mode when system enters this "PM2" mode (associated to suspend in zephyr). In this case, the UART is not able to really do any send and transceive normally as far as I could tell but it could change clock source to some really slow like 1Hz clock which is the only one left active on the system in this mode, and be able to "receive" one bit/byte, and this activity can then be a wakeup cause of the chip. So what this driver is essentially saying is "if my normal function is going to be disabled, reconfigure my clock to the low power clock and configure me as wake up source" (zephyr,disabling-states). Whereas on "PM3" mode (associated to suspend), the UART just completely lose power and state and cannot be a wakeup. So we set the power-domain to this "PM3" domain which is currently just a stupid mechanism to be able to deliver an "OFF" action to all these devices which lose power and state from "standby" mode, and set "zephyr,disabling-states" as "suspend" (and standby technically, but irrelevant) so that it knows it loses function but can reconfigure itself as a wakeup...
> 
> So do you have any idea about how to handle this problem which is that our device cannot only just be simply "disabled" or "enabled" but can also be in this type of low power mode that it needs to account for? Because from PM device callbacks alone, it's not enough here to handle that "PM2" case, we can handle this PM3 case to get the OFF action and know we will lose state but I am not sure how to differentiate for the "low power operation" case

Any part of the system, be that drivers, subsystems and applications, can register themselves using `pm_notifier_register()` to be able to follow the soc sleep states as they are entered and exited. devicetree nodes can reference these states freely. For convenience, for devices which can only be enabled/disabled as a whole depending on soc sleep states, can use the default `zephyr,disabling-states`. Devices which are affected in a more granular way, can define their own sets of states as they see fit, like `transmit-disabling-states` in the case of this specific UART. These are device specific, and the device driver can lock/unlock/monitor these states freely.

Commit:
pm: Extend pm notifier to be able to report substate

I am surprised this notifier didn't already report the substate id, it
seems important since different substate obviously are defined for a
reason, they can be having a different effect on the system.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
pm: Add constraint lists and APIs

Some pieces of the system may have custom types of constraints that they
define based on different effects and reasons than just the standard
"zephyr,disabling-states". To avoid every single one of these component
reinventing the wheel, make some common APIs to handle these type of
custom constraint lists.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: uart_mcux_flexcomm: Refactor low power state handle

Refactor low power state handling to not tie to zephyr,disabling-states
and define it's own separate "low power states" property in DT instead.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

bjarki-andreasen commented on include/zephyr/pm/policy.h:
```diff
@@ -125,6 +125,9 @@ void pm_policy_state_lock_get(enum pm_state state, uint8_t substate_id);
  */
 void pm_policy_state_lock_put(enum pm_state state, uint8_t substate_id);
 
+void pm_policy_state_lock_get_list(struct pm_state_constraint_list* constraints);
+void pm_policy_state_lock_put_list(struct pm_state_constraint_list* constraints);
```
Since we are not actually passing a "list" of states `(enum pm_state *states, uint8_t *substate_ids, uint8_t size)` we could instead name the function according to the type passed to it:
```
void pm_policy_state_constraint_list_lock_get()
void pm_policy_state_constraint_list_lock_get()
```
and also, the name `struct pm_state_constraint_list` could be `struct pm_state_constraints`, which I think looks nicer, and sounds better if used in the new APIs
```
void pm_policy_state_constraints_lock_get()
```
and even remove the `lock` part:
```
void pm_policy_state_constraints_get()
```
later we can replace uses of `pm_policy_state_lock_put` since its the same but with a "list" of size 1

bjarki-andreasen commented on include/zephyr/pm/state.h:
```diff
@@ -358,6 +364,34 @@ struct pm_state_constraint {
+	PM_STATE_CONSTRAINT_INIT(DT_PHANDLE_BY_IDX(node_id, phandle, idx))
+
+#define PM_STATE_CONSTRAINTS_LIST_NAME(node_id, phandles)                     \
+	node_id ## _ ## phandles ## _constraints
```
use `_CONCAT_4()` to avoid preprocessor issues with macro arg expansions

bjarki-andreasen replied:
Also prepend with Z_ for internal macros

bjarki-andreasen commented on include/zephyr/pm/state.h:
```diff
@@ -358,6 +364,34 @@ struct pm_state_constraint {
+#define PM_STATE_CONSTRAINTS_LIST_NAME(node_id, phandles)                     \
+	node_id ## _ ## phandles ## _constraints
+
+#define PM_STATE_CONSTRAINTS_LIST_DEFINE(node_id, phandles)                                \
```
replace `phandles` with `prop`, the user does not care what type the prop is when using a macro like this, just the name of the prop. Also add a bit of documentation to it

bjarki-andreasen commented on include/zephyr/pm/state.h:
```diff
@@ -358,6 +364,34 @@ struct pm_state_constraint {
+		.substate_id = DT_PROP_OR(node_id, substate_id, 0),           \
+	}
+
+#define PM_STATE_CONSTRAINT_REF(node_id, phandle, idx)                        \
```
Prepend with Z_ for internal macros

bjarki-andreasen commented on include/zephyr/pm/state.h:
```diff
@@ -358,6 +364,34 @@ struct pm_state_constraint {
+ *
+ * @param node_id Node identifier.
+ */
+#define PM_STATE_CONSTRAINT_INIT(node_id)                                     \
```
Prepend with Z_ for internal macros

bjarki-andreasen commented on include/zephyr/pm/state.h:
```diff
@@ -358,6 +364,34 @@ struct pm_state_constraint {
+		DT_FOREACH_PROP_ELEM_SEP(node_id, phandles, PM_STATE_CONSTRAINT_REF, (,))  \
+	}
+
+#define PM_STATE_CONSTRAINTS_LIST_GET(node_id, phandles)                                   \
```
Following the previous note about the naming, `PM_STATE_CONSTRAINTS_GET` would be nicer IMO

bjarki-andreasen commented on subsys/pm/state.c:
```diff
@@ -107,3 +107,20 @@ const struct pm_state_info *pm_state_get(uint8_t cpu, enum pm_state state, uint8
 	return NULL;
 }
+
+bool pm_state_in_constraints(const struct pm_state_constraint_list *constraints,
```
like this! sounds way nicer than `pm_state_in_constraint_list`

bjarki-andreasen reviewed (changes requested):
This approach is really nice! I would not say this is a temporary approach, the approach in this PR is a way more scalable variant of the pm state lock get and puts we have currently, which can replace the current pm_policy_device_power_lock_get() (which we still need to call from power domains anyway). Added notes mostly about naming

decsny commented:
rebased & resolve conflicts

bjarki-andreasen reviewed (dismissed):
Really nice, I think this is the way to go forward

Commit:
tests: pm: Add constraint list API test

Add test for lock/unlock constraint lists using the new arbitrary DT
phandles macros and pm_policy_state_constraints_get/put APIs.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>
